### PR TITLE
Process cells

### PR DESCRIPTION
**Pull request by ErinWeisbart**
Here's the draft of 2.process-cells.
I *think* it's good to go through line 132.
After line 133  # Add metadata is where I need the major help.

(Also, I haven't run it through black yet.)

### CODE REVIEW

File: 0.preprocess-sites/2.process-cells.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        print(f"{site} data not found")
+        continue
+
+
+    # Merge compartment csvs
+    sc_merged_df = merge_single_cell_compartments(compartment_csvs, merge_info, id_cols)
+    sc_merged_df = sc_merged_df.sort_values(by=sort_col)
+
+    # Add metadata
+    sc_merged_df = sc_merged_df.rename({"Metadata_Cells_ImageNumber": "Metadata_ImageNumber"}, axis="columns")
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:134]: This rename step is necessary for the following merge. 
I wasn't sure the right way to extract the hardcoded column names out?

File: 0.preprocess-sites/2.process-cells.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    sc_merged_with_metadata_df = foci_df.merge(
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:135]: This merge is as written for CP074B with the merge columns extracted.
But I think something is wrong because the resulting sc_merged_with_metadata_df has a LOT of NaN.

File: 0.preprocess-sites/2.process-cells.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        sc_merged_df,
+        left_on=metadata_merge_left,
+        right_on=metadata_merge_right,
+        how="right",
+    )
+
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:141]: I haven't yet tested anything below this point because the sc_merged_with_metdata_df looked wrong to me.

File: 0.preprocess-sites/2.process-cells.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    metadata_df = sc_merged_with_metadata_df.loc[
+        :, sc_merged_with_metadata_df.columns.str.contains("Metadata")
+    ].drop_duplicates()
+
+    metadata_df.loc[
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:146]: I don't understand what these rows 146-152 are doing so I couldn't extract them out.

File: scripts/config_utils.py [hunk 5]
```diff
@@ -47,6 +47,13 @@ def preprocess_sites_config(config):
         config["process-spots"]["output_basedir"], config["core"]["batch"], "spots"
     )
 
+    # Build paths to the foci files
```
Comment (ErinWeisbart) [inline scripts/config_utils.py:50]: We don't actually need this since we added `foci_dir = spot_args["output_spotdir"]` to 2.process-cells.py instead. To delete this addition do I just make another commit where I've deleted this section and add that commit to the PR?  
 
Sidenote: At least I'm not going totally crazy because I knew I had added this but somehow totally missed it when we were adding in the call to the spot_args!

Comment (gwaygenomics) [reply scripts/config_utils.py:50]: > To delete this addition do I just make another commit where I've deleted this section and add that commit to the PR?

Yep, exactly

> Sidenote: At least I'm not going totally crazy because I knew I had added this but somehow totally missed it when we were adding in the call to the spot_args!

Great detective work! 🕵️

File: 0.preprocess-sites/2.process-cells.py (hunk 1 above)
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:134]: I like the strategy in https://github.com/broadinstitute/pooled-cell-painting-profiling-recipe/pull/13/files#diff-d7c74c21a3b3d2924816aa921e827762R80

If we change it to `Metadata_Cells_ImageNumber` then a renaming isn't necessary.

File: 0.preprocess-sites/2.process-cells.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+prefilter_dir = prefilter_args["output_basedir"]
+prefilter_file = prefilter_args["prefilter_file"]
+
+foci_dir = spot_args["output_spotdir"]
+
+sort_col = cell_args["sort_col"]
+output_basedir = cell_args["output_basedir"]
+merge_info = cell_args["merge_columns"]
+metadata_merge_left = cell_args["metadata_merge_columns"]["left"]
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:69]: love this strategy! Can we make the `left` and `right` more descriptive somehow?

File: 0.preprocess-sites/2.process-cells.py (hunk 2 above)
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:135]: we expect a certain amount of NAs. The key line is [line 139](https://github.com/broadinstitute/pooled-cell-painting-profiling-recipe/pull/13/files#diff-60750dd97846eb5d02654ce784c9eb1fR139) `how="right"`.

This means that the dataframe on the right (`sc_merged_df`) (as opposed to the left, `foci_df`) is how we decide to introduce NAs. Biologically, if there are cells in `sc_merged_df` (based on the `metadta_merge_left` columns) **without foci** then all of the foci columns will have missing values (aka NAs).

[`pd.merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) is pretty tricky, but massively powerful.

Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:135]: do we expect CP151 to have many unassigned cells?

File: 0.preprocess-sites/2.process-cells.py (hunk 3 above)
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:141]: is there a way we could check to make sure its not?

File: 0.preprocess-sites/2.process-cells.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        right_on=metadata_merge_right,
+        how="right",
+    )
+
+    metadata_df = sc_merged_with_metadata_df.loc[
+        :, sc_merged_with_metadata_df.columns.str.contains("Metadata")
+    ].drop_duplicates()
+
+    metadata_df.loc[
+    metadata_df.Metadata_Foci_Cell_Category.isna(), "Metadata_Foci_Cell_Category"
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:147]: this is a quick one - this is asking: how should a cell without foci be categorized?

File: 0.preprocess-sites/2.process-cells.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    ] = 5
+    metadata_df.Metadata_Foci_Cell_Category = metadata_df.Metadata_Foci_Cell_Category.astype(
+        int
+    )
+    metadata_df.loc[metadata_df.Metadata_Foci_site.isna(), "Metadata_Foci_site"] = site
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:152]: `Metadata_Foci_site` is NA when cells don't have foci. But the site still exists and shouldn't be NA.

File: 0.preprocess-sites/2.process-cells.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+    metadata_df = sc_merged_with_metadata_df.loc[
+        :, sc_merged_with_metadata_df.columns.str.contains("Metadata")
+    ].drop_duplicates()
+
+    metadata_df.loc[
+    metadata_df.Metadata_Foci_Cell_Category.isna(), "Metadata_Foci_Cell_Category"
+    ] = 5
+    metadata_df.Metadata_Foci_Cell_Category = metadata_df.Metadata_Foci_Cell_Category.astype(
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:149]: this decision is going to need some more thought... If `Metadata_Foci_Cell_Category` is ever _not_ an int, this will fail out.

File: 0.preprocess-sites/site_processing_config.yaml [hunk 10]
```diff
@@ -61,3 +63,20 @@ process-cells:
...
+  merge_columns:
+    image_column: ImageNumber
+    linking_compartment: Cytoplasm
+    linking_columns:
+      cells: Metadata_Cytoplasm_Parent_Cells
+      nuclei: Metadata_Cytoplasm_Parent_Nuclei
+    metadata_linking_columns:
+      - Metadata_Foci_site
+      - Metadata_Cells_ObjectNumber
+  metadata_merge_columns:
```
Comment (gwaygenomics) [inline 0.preprocess-sites/site_processing_config.yaml:75]: 🎉

File: 0.preprocess-sites/2.process-cells.py (hunk 6 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:69]: yes! I just don't understand how we choose the columns for `left` and `right` well enough to have a more descriptive name. you can help me with that :)

File: 0.preprocess-sites/2.process-cells.py (hunk 2 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:135]: By unassigned cells you mean cells without foci?
I believe that is something that will be consistent between experiments because it's based on the MOI (infection rate) and that's not something that is likely to change in the immediate future.

File: 0.preprocess-sites/2.process-cells.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+sites = [x for x in os.listdir(batch_dir) if x not in ignore_files]
+
+for site in sites:
+    try:
+        print(f"Now processing {site}...")
+        output_folder = pathlib.Path(output_basedir, batch, "paint", site)
+        os.makedirs(output_folder, exist_ok=True)
+
+        meta_output_file = pathlib.Path(output_folder, f"metadata_{site}.tsv.gz")
```
[the remark this answers is not in the document]
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:96]: Notes:
If we make a new PR that ever touches these files (they are not currently touched by anything downstream) we can update the meta_output_file in the PR so that nothing can break.

File: 0.preprocess-sites/2.process-cells.py (hunk 7 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:146]: It needs to be an int that is +1 the number of cell categories (at least for now). This will change when we move away from simple categories.

File: 0.preprocess-sites/2.process-cells.py (hunk 9 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:148]: (so we'll delete line 148 and instead move to defining the type when it's written out to csv)

File: 0.preprocess-sites/2.process-cells.py (hunk 6 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:69]: After discussion, we'll rename metadata_merge_left to metadata_merge_foci_cols and metadata_merge_right to metadata_merge_cell_cols.

File: 0.preprocess-sites/2.process-cells.py (hunk 2 above)
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:134]: Note: I was misreading the df. The NaN are expected. The NaN are populating the foci metadata for cells that do not have foci in them.

File: 0.preprocess-sites/2.process-cells.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ErinWeisbart) [reply 0.preprocess-sites/2.process-cells.py:95]: These are actually both loaded by 7.visualize-cell-summary so we will need to move these to config when we re-write that script.

File: 0.preprocess-sites/2.process-cells.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+            cell_category_df,
+            left_on=metadata_cell_quality_col,
+            right_index=True,
+            how="left",
+        )
+        .sort_values(by=cell_sort_col)
+        .reset_index(drop=True)
+    )
+
+    # Create a summary of counts of each cell quality class
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:168]: This is a completely different way than you were making this df. I had to make some major changes thanks to deprecations in pandas 1.0. Hopefully I didn't simplify this too much in the process and miss something from your method?

**Comment by ErinWeisbart** [comment]
I think it's now fully functional and ready for a real review!
The only thing that I haven't addressed in comments is that the final metadata_site.tsv doesn't look exactly the same as the output for CP074B. It has columns that were dropped in CP074B. The list of columns kept was hardcoded in CP074B. It's not a lot of extra columns so it didn't seem worth it to me to introduce a list that would have to be edited in config, especially as one can instead drop it to the necessary columns at the start of future manipulation. But I certainly defer to you on this point.

File: 0.preprocess-sites/2.process-cells.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        how="right",
+    )
+
+    # Drops all columns that are not Metadata (i.e. profiles)
+    metadata_df = sc_merged_with_metadata_df.loc[
+        :, sc_merged_with_metadata_df.columns.str.contains("Metadata")
+    ].drop_duplicates()
+
+    # Adds a cell quality category to previously uncategorized cells
+    metadata_df[metadata_cell_quality_col] = metadata_df[
```
Comment (ErinWeisbart) [inline 0.preprocess-sites/2.process-cells.py:149]: @gwaygenomics These .fillna are a different approach than you were doing for CP074B. (You had used .isna but I couldn't figure out how to remove the hardcoded column name from your method). Let me know if you see a reason to go back to your method that I have missed.

File: 0.preprocess-sites/2.process-cells.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+merge_info = cell_args["merge_columns"]
+metadata_merge_foci_cols = cell_args["metadata_merge_columns"]["foci_cols"]
+metadata_merge_cell_cols = cell_args["metadata_merge_columns"]["cell_cols"]
+metadata_cell_quality_col = cell_args["metadata_merge_columns"]["cell_quality_col"]
+foci_site_col = cell_args["foci_site_col"]
+
+cell_quality = CellQuality(quality_func)
+cell_category_dict = cell_quality.define_cell_quality()
+empty_cell_category = len(cell_category_dict) + 1
+cell_category_dict[empty_cell_category] = "Empty"
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:73]: confirmed that we've added a reference in https://github.com/broadinstitute/pooled-cell-painting-profiling-recipe/issues/10#issuecomment-636196358

File: 0.preprocess-sites/2.process-cells.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:95]: great - it is ok to keep in this PR, we just need to remember to update later.

File: 0.preprocess-sites/2.process-cells.py (hunk 13 above)
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:149]:
```suggestion
    metadata_df.loc[:, metadata_cell_quality_col] = metadata_df.loc[:, 
```

Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:149]: does this work? In general, I much prefer making the `.loc` call and referencing the column, than the approach without making it explicit that we are referencing a column. If you agree, then we'll likely need to rerun `black` after committing the update.

File: 0.preprocess-sites/2.process-cells.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+        )
+        .sort_values(by=cell_sort_col)
+        .reset_index(drop=True)
+    )
+
+    # Create a summary of counts of each cell quality class
+    quality_counts = metadata_df.Cell_Class.value_counts()
+    cell_count_df = pd.DataFrame(quality_counts)
+    cell_count_df = cell_count_df.rename(columns={"Cell_Class": "cell_count"})
+    cell_count_df["site"] = site
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:172]:
```suggestion
    cell_count_df = cell_count_df.assign(site=site)
```

lets switch to [`pd.assign()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.assign.html).

File: 0.preprocess-sites/2.process-cells.py (hunk 12 above)
Comment (gwaygenomics) [reply 0.preprocess-sites/2.process-cells.py:168]: this looks great! Thanks for catching the deprecations.

btw, if inclined, we can switch this to a nice pandas chain ⛓️:

```python
cell_count_df = pd.DataFrame(
  metadata_df
  .Cell_Class
  .value_counts()
  .rename(columns={"Cell_Class": "cell_count"})
  .assign(site=site)
)
```

File: 0.preprocess-sites/2.process-cells.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+    # Save files
+    metadata_df.to_csv(meta_output_file, sep="\t", index=False)
+    cell_count_df.to_csv(count_output_file, sep="\t", index_label="Cell_Quality")
+
+print("All sites complete.")
```
Comment (gwaygenomics) [inline 0.preprocess-sites/2.process-cells.py:178]: 🎉
